ERCS/erc-7579.md

```diff
@@ -48,149 +48,164 @@ The smart account's `validateUserOp` function SHOULD return the return value of
 
 #### Execution Behavior
 
-To comply with this standard, smart accounts MUST implement the two interfaces below. If an account implementation elects to not support any of the execution methods, it MUST implement the function, but revert instead of complying with the specifications below. This is in order to avoid unpredictable behavior with fallbacks.
+To comply with this standard, smart accounts MUST implement the execution interface (except for the optional function) below:
 
 ```solidity
 interface IExecution {
     /**
-    * @dev packing batched transactions in a struct saves gas on both ERC-4337 and Executor Module flows
-    */
-    struct Execution {
-        address target;
-        uint256 value;
-        bytes callData;
-    }
+     * @dev Executes a transaction on behalf of the account.
+     *         This function is intended to be called by ERC-4337 EntryPoint.sol
+     * @param mode The encoded execution mode of the transaction. See ModeLib.sol for details
+     * @param executionCalldata The return data of the executed contract call
+     *
+     * MUST implement this function
+     * MUST ensure adequate authorization control: i.e. onlyEntryPointOrSelf
+     * If a mode is requested that is not supported by the Account, it MUST revert
+     */
+    function execute(bytes32 mode, bytes calldata executionCalldata) external payable;
```

Comment:
Eg see [this example](https://github.com/erc7579/uMSA/blob/07ad177612c241ad6703b7a818cdda3204b7add5/test/MSA.t.sol#L22)

Reply:
the normal call flow with current EP:
ep->account.execute -> target.calldata

but with your suggested API, the flow becomes:
ep->account.execute ->account.callTarget -> target.calldata
 
which is an extra call level (or actually, 2 extra call levels, since each call into the account goes through a proxy)

Reply:
No it is the normal flow but instead of passing `(address target, uint256 value, bytes calldata data` to the execution function we just pass a `bytes` object that has these params encoded (so that the same bytes can be used for single and batched executions

Reply:
eg for a single execution you could do `bytes executionCalldata = abi.encodePacked(target, value, data)`

Reply:
Yeah you're right, the tradeoff is on the spectrum between a single function for everything and a specific function for everything. In the previous draft, we had different execution function for every permutation of execution mode (eg call/delegatecall, single/batch, fromEP/fromExecutor, revertOnFailure/try, ...). 

This obviously scales exponentially with every execution mode that is added, so we made the decision to restrict the execution functions to execute and executeFromExecutor, because these are fundamentally different from a security perspective. To further encompass execution differences, we introduced modes in order to give account builders the flexibility to innovate on how the execution on their account works. 

To some extent where we drew the line is arbitrary but implementing every execution permutation as its own function is not future proof at all.